burnman/layer.py

```diff
@@ -0,0 +1,797 @@
+# Copyright (C) 2012 - 2017 by the BurnMan team, released under the GNU
+# GPL v2 or later.
+
+import numpy as np
+from scipy.integrate import odeint
+from scipy.integrate import quad
+from scipy.interpolate import UnivariateSpline
+from burnman import averaging_schemes
+from burnman import constants
+from burnman import geotherm
+import warnings
+
+
+from .material import Material, material_property
+from .mineral import Mineral
+from .composite import Composite
+from .seismic import Seismic1DModel
+
+import matplotlib.pyplot as plt
+
+
+class Layer(object):
+    """
+    A planetary layer class
+    """
+
+    def __init__(self, name=None, radius_planet=None,
+                 min_depth=None, max_depth=None, n_slices=None, verbose=False):
+
+        self.name = name
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+        self.thickness = self.max_depth - self.min_depth
+        self.n_slices = n_slices
+        self.depths = np.linspace(min_depth, max_depth, n_slices)
+        self.radius_planet = radius_planet
+        self.radii = self.radius_planet - self.depths
+        self.outer_radius = max(self.radii)
+        self.inner_radius = min(self.radii)
+        self.verbose = verbose
+        self._cached = {}
+        self._pressures = None
+        self._temperatures = None
+        self.sublayers = None
+    
+    def reset(self):
+        """
+        Resets all cached material properties.
+
+        It is typically not required for the user to call this function.
+        """
+        self._cached = {}
+        self._pressures = None
+        self._temperatures = None
+        self.sublayers = None
+
+    def set_material(self, material):
+        """
+        Set the material of a Layer with a Material
+        """
+        assert(isinstance(material, Material))
+        self.material = material
+        self.reset()
+
+    def set_temperature_mode(self, temperature_mode='adiabat',
+                             temperatures=None, temperature_top=None):
+        """
+        Sets temperature of the layer by user-defined values or as an (modified) adiabat.
+        temperature_mode is 'user_defined','adiabatic', or 'modified_adiabat'
+
+        Parameters
+        ----------
+        temperature_mode : string
+        This can be set to 'user_defined','adiabatic', or 'modified_adiabat'
+        temperatures : array of float
+        The desired fixed temperatures in [K]. Should have same length as defined depths in layer.
+        temperature_top : float
+        Temperature at the top for an adiabat
+        
+        Note
+        ---------
+        'user_defined' = fixes the temperature with the profile input by the users
+        'adiabat' = self-consistently computes the adiabat when setting the state of the layer
+        'modified_adiabat' = adds the user input array to the adiabat, 
+            e.g. allows to implement boundary layers
```

Comment:
I like this. We should also consider implementing the McKenzie (2005) boundary layer model (or a more modern alternative). I've been using it for Mars, and I think it would be handy for others as well.